frodo.js

```diff
@@ -1,39 +1,37 @@
-/**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?